Add pool diagnostics to the internal transport and agent-core

tiagolobocastro opened the pull request:
refactor(pstor/pool): split pool spec types
    
    As with other types, the pool pstor info was incorrectly stored as the spec.
    This makes it harder to separate user spec and other information.
    
    For now we simply create an innter type PoolUSpec which is basically the user
    specification of a pool.
    This allows us to transport pool spec and pool diag separately.
    
---

    test(pool/diag): add pool diag info check to pool reconciler
    
    Tests both disk not found and invalid superblock.
    
---

    feat(pool/diag): collect import errors during pool import
    
    When pool import fails, collect error reason.
    This is stored in runtime information within the PoolSpec.
    
    A downside of this is that the PoolSpec in the pstor, alike the volume,
    is now a misnomer.
    In the future we should introduce an upgrade way of updating the entries.
    
---

    feat(pool/diag): add grpc proto and transport interface

tiagolobocastro committed:
feat(pool/diag): add grpc proto and transport interface

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
feat(pool/diag): collect import errors during pool import

When pool import fails, collect error reason.
This is stored in runtime information within the PoolSpec.

A downside of this is that the PoolSpec in the pstor, alike the volume,
is now a misnomer.
In the future we should introduce an upgrade way of updating the entries.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro commented on control-plane/grpc/proto/v1/pool/pool.proto:
```diff
@@ -5,6 +5,36 @@ import "google/protobuf/wrappers.proto";
 
 package v1.pool;
 
+message PoolHealth {
```
```suggestion
message PoolDiagnostics {
```

Wondering if this would be a better naming?

abhilashshetty04 replied:
This will always have import errors in them. If yes, we can just call it as ImportError.

struct ImportError {
errors: Vec<DiskPoolError>
}

tiagolobocastro replied:
Not necessarily, this may be expanded for something like keep alives on a timer, smart data etc

Abhinandan-Purkait replied:
Actually it's a difficult choice here, the process is diagnostic but the outcome comes as a health metric. So, maybe health makes sense?

tiagolobocastro replied:
Yeah it is related to health, but it's not exactly the pool health, this doesn't tell you if the pool is healthy, only reports diagnostic information?

abhilashshetty04 replied:
I think we can keep it as Diagnostics.

abhilashshetty04 commented on control-plane/agents/src/bin/core/pool/operations_helper.rs:
```diff
@@ -60,6 +60,56 @@ impl OperationGuardArc<PoolSpec> {
+            match error.tonic_code() {
+                tonic::Code::NotFound | tonic::Code::InvalidArgument => {
+                    let disks = pool_spec.disks.first().map(|d| d.to_string());
+                    *reporter.lock().expect("not poisoned") = Some(PoolHealth {
```
Will we clear this if import passes eventually?

tiagolobocastro replied:
Yes that's already here on the success complete case

tiagolobocastro commented:
> @tiagolobocastro, Can we write unit test for this?

I tried but it's tricky, when a device is missing on container restarts it creates a directory in its place
I could try with a file instead of a block device

abhilashshetty04 commented:
nit: commits have different type.. feat(pool/health) and feat(pool-health) was it intended?

abhilashshetty04 commented on control-plane/stor-port/src/types/v0/transport/pool.rs:
```diff
@@ -80,6 +80,52 @@ impl Deref for CtrlPoolState {
+#[serde(rename_all = "camelCase")]
+pub struct PoolHealth {
+    /// Errors encountered when trying to import the pool.
+    pub import_errors: Vec<PoolDiskError>,
```
Does it need to be vector? Are we going to store more then 1 errors or just the latest failure information.

tiagolobocastro replied:
In case there's more than 1 error detected.

abhilashshetty04 replied:
Suppose we face error during request no 1: DiskNotFound. During request no 3 Disk was there but failed due to crc mismatch. So will we have both of them in import errors vec?

Is it possible that we may see more then 1 error in one request?

tiagolobocastro replied:
On the same request; say you have 2 disks, 1 not found, another is ReadIoError;
Also in the future we may have other tests which we perform before import, ex: latency tests, etc

tiagolobocastro committed:
test(pool/diag): add pool diag info check to pool reconciler

Tests both disk not found and invalid superblock.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

abhilashshetty04 commented:
> > > @tiagolobocastro, Can we write unit test for this?
> > 
> > 
> > I tried but it's tricky, when a device is missing on container restarts it creates a directory in its place I could try with a file instead of a block device
> 
> @abhilashshetty04 the reconciler tests were already doing something similar to enhanced that to check for invalid super block and not found errors.

Is it there on the same PR?

tiagolobocastro commented:
> > > > @tiagolobocastro, Can we write unit test for this?
> > > 
> > > 
> > > I tried but it's tricky, when a device is missing on container restarts it creates a directory in its place I could try with a file instead of a block device
> > 
> > 
> > @abhilashshetty04 the reconciler tests were already doing something similar to enhanced that to check for invalid super block and not found errors.
> 
> Is it there on the same PR?

Yes it's here

Abhinandan-Purkait commented on control-plane/agents/src/bin/core/pool/operations_helper.rs:
```diff
@@ -60,6 +60,56 @@ impl OperationGuardArc<PoolSpec> {
+            match error.tonic_code() {
+                tonic::Code::NotFound | tonic::Code::InvalidArgument => {
+                    let disks = pool_spec.disks.first().map(|d| d.to_string());
+                    *reporter.lock().expect("not poisoned") = Some(PoolHealth {
```
Do we need the mutex here? I couldn't spot any critical section here?

tiagolobocastro replied:
Yes we need mutex because the reporter is shared, so we need mutex to get a mutable reference

tiagolobocastro committed:
refactor(pstor/pool): split pool spec types

As with other types, the pool pstor info was incorrectly stored as the spec.
This makes it harder to separate user spec and other information.

For now we simply create an innter type PoolUSpec which is basically the user
specification of a pool.
This allows us to transport pool spec and pool diag separately.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>
